### DISCUSSION TITLE
How to Initialize a Provider with asynchronous values?

### DISCUSSION THREAD

**Discussion by Helzio** [rrousselGit/riverpod#614] [Q&A]
In my app I'm using Dio for http request. I have a `dioProvider` like this:
```dart
final dioProvider = Provider<Dio>((ref) => Dio());
```

Later, I try to add a plugin to Dio to manage the cookies but a can't use `async` from a "normal" `Provider`:
```dart
// This code does not compile
final dioProvider = Provider<Dio>((ref) async{
  final dio = Dio();
  final appDocDir = await getApplicationDocumentsDirectory();
    final appDocPath = appDocDir.path;
    dio.interceptors.add(
      CookieManager(
        PersistCookieJar(
          ignoreExpires: true,
          storage: FileStorage("$appDocPath/.cookies/"),
        ),
      ),
    );
  return dio;
});
```

I see this approach of `SharedPreferences` throwing an `UnimplementedError`, waiting for the `SharedPreferences` instance in the main method and later overwriting the `Provider` value inside the `ProviderScope`:
```dart
final sharedPreferencesProvider = Provider<SharedPreferences>((_) => throw UnimplementedError());

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  final sharedPreferences = await SharedPreferences.getInstance();
  runApp(
    ProviderScope(
      overrides: [
        sharedPreferencesProvider.overrideWithValue(sharedPreferences),
      ],
      child: MyApp(),
    ),
  );
}

```

But it may be the case that the value that we have to wait to initialize a provider takes some time and we don't want to block the start of the application.

My question is:
**How to overwrite providers with asynchronous values until we need them but not awaiting for the values in the main method?**

For example. Returning to the example of the `dioProvider`. I have a `loginRepositoryProvider` that depends on it, and a `loginNotifierProvider` (`StateNotifierProvider`) that depends on the repository...
What would be the exact way and time to initialize the `dioProvider` so that everything works correctly?

I hope you can help me. Thanks for taking the time to read my question.

**Comment by Moussenger** [comment]
Why can not dioProvider be a `FutureProvider`?

Then, you only have to await for that future in the providers that use it.
